gateway-js/src/index.ts

```diff
@@ -467,6 +445,102 @@ export class ApolloGateway implements GraphQLService {
+    const previousCsdl = this.parsedCsdl;
+    const previousCompositionId = this.compositionId;
+
+    if (previousSchema) {
+      this.logger.info('Updated CSDL was found.');
+    }
+
+    await this.maybePerformServiceHealthCheck(result);
+
+    this.compositionId = result.id;
+    this.parsedCsdl = parse(result.csdl);
+
+    if (this.queryPlanStore) this.queryPlanStore.flush();
+
+    const { schema, composedSdl } = this.createSchema({
+      csdl: result.csdl,
+    });
+
+    if (!composedSdl) {
+      this.logger.error(
+        "A valid schema couldn't be composed. Falling back to previous schema.",
+      );
+    } else {
+      this.schema = schema;
+      this.queryPlannerPointer = getQueryPlanner(composedSdl);
+
+      // Notify the schema listeners of the updated schema
+      try {
+        this.onSchemaChangeListeners.forEach((listener) =>
+          listener(this.schema!),
+        );
+      } catch (e) {
+        this.logger.error(
+          "An error was thrown from an 'onSchemaChange' listener. " +
+            'The schema will still update: ' +
+            ((e && e.message) || e),
+        );
+      }
+
+      if (this.experimental_didUpdateComposition) {
+        this.experimental_didUpdateComposition(
+          {
+            compositionId: result.id,
+            csdl: result.csdl,
+            schema: this.schema,
+          },
+          previousCompositionId && previousCsdl && previousSchema
+            ? {
+                compositionId: previousCompositionId,
+                csdl: print(previousCsdl),
+                schema: previousSchema,
+              }
+            : undefined,
+        );
+      }
+    }
+  }
+
+  private async maybePerformServiceHealthCheck(update: CompositionUpdate) {
+    // Run service health checks before we commit and update the new schema.
+    // This is the last chance to bail out of a schema update.
+    if (this.config.serviceHealthCheck) {
+      const serviceList = isCsdlUpdate(update)
+        ? this.serviceListFromCsdl(parse(update.csdl))
+        : // Existence of this is determined in advance with an early return otherwise
+          update.serviceDefinitions!;
+      // Here we need to construct new datasources based on the new schema info
+      // so we can check the health of the services we're _updating to_.
+      const serviceMap = serviceList.reduce((serviceMap, serviceDef) => {
+        serviceMap[serviceDef.name] = {
+          url: serviceDef.url,
+          dataSource: this.createDataSource(serviceDef),
+        };
+        return serviceMap;
+      }, Object.create(null) as DataSourceMap);
+
+      try {
+        await this.serviceHealthCheck(serviceMap);
+      } catch (e) {
+        this.logger.error(
+          'The gateway did not update its schema due to failed service health checks.  ' +
+            'The gateway will continue to operate with the previous schema and reattempt updates.' +
+            e,
+        );
+        throw e;
```

Comment:
While it's not new here, this is a case of the odd pattern of "log and rethrow". In general having any given error only handled in one place. This error will already be either logged in pollServices or thrown from load.  Why not augment the error with the extra text instead of logging here? That way you don't end up with the error logged twice (or logged despite a throw from `load` being caught).

Reply:
Agree. Let me know what you think about this change 692e046. At first I removed the original error altogether, but I think with the addition of related service name that failed the health check, the original error is now much more useful. Updated snapshots tell the story.